rfc/0006-kubewarden-versioning.md

```diff
@@ -0,0 +1,203 @@
+    - `docker.json`
+    - sigstore related files
+
+There is also the `waPC` contract between the `policy-evaluator`
+(consumed by the Policy Server and kwctl) and Policies and SDK's.
+
+## Kubewarden toplevel version
+
+This proposal envisions using [`CalVer`](https://calver.org/) to
+version the Kubewarden project.
+
+Had Kubewarden be released today, it would be `Kubewarden
+22.03`. This makes it trivial to understand how old your version of
+Kubewarden is, and how much your installation is drifting from the
+latest releases.
+
+## Kubewarden component versions
+
+Every component in Kubewarden might have its own versioning. The
+Kubewarden maintainers are going to maintain a list of all releases.
+
+This list of releases looks as follows:
+
+```yaml
+"22.03":
+  - component: kubewarden-crds-chart
+    version: 0.1.1
+  - component: kubewarden-controller-chart
+    version: 0.3.6
+  - component: kubewarden-controller
+    version: v0.4.5
+  - component: policy-server
+    version: v0.2.6
+  - component: kwctl
+    version: v0.2.5
+"22.05":
+  - component: kubewarden-crds-chart
+    version: 0.2.0
+  - component: kubewarden-controller-chart
+    version: 0.4.0
+  - component: kubewarden-controller
+    version: v0.4.8
+  - component: policy-server
+    version: v0.3.0
+  - component: kwctl
+    version: v0.3.1
+...
+```
+
+As more components are relevant they can be added to the list as time
+grows.
+
+This map allows us to generate documentation and a single source of
+truth to refer to when a user has a problem, or when we are debugging
+an issue. It helps to understand if any component has drifted from the
+version we are expecting.
+
+The helm chart will be responsible of storing the "Kubewarden version"
+in a resource, like a ConfigMap. So it's easy to inspect it and
+understand what "Kubewarden version" a cluster is running.
+
+Had other external components such as `kwctl` need to synchronize with
+the cluster, they can inspect this ConfigMap and realize if they are
+compatible with the running version of Kubewarden on the cluster they
+are connecting to.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- It's not a very common versioning scheme. I would consider reading
+  ["When to use
+  CalVer"](https://calver.org/#when-to-use-calver). Although
+  Kubewarden does not have a "large" scope, it's to some extent
+  "constantly changing", and this constant change would imply
+  harmonizing versions of all components continuously.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Have a regular Kubewarden toplevel version
+
+Having a regular --even semver based-- version on the toplevel project
+seems like an artificial construct. What is the meaning of patchlevel
+release in this case? What is Kubewarden 1.2.3 vs 1.2.4?
+
+Thus, ignoring patchlevel (e.g. just keeping 1.2 vs 1.3 vs 1.4, or
+even 1 vs 2 vs 3) is an option. But then, if it's just meaningless
+numbers, why not make the toplevel number meaningful, referring to the
```

Comment:
to me with a user hat, having major, minor, patch would inform me that I can upgrade without headache and needing to reconfigure things.
No matter the changes in the policy-server image, if I can just bump it and rollout the deployment without backwards incompatible changes, that's a bump in minor or patch version. Same for other components. Now, if I want to benefit from new features (bump in minor), then, I may need to enable them by hand, or maybe the come enabled by default.

Reply:
> No matter the changes in the policy-server image, if I can just bump it and rollout the deployment without backwards incompatible changes, that's a bump in minor or patch version

Yeah, that's why I started this RFC with the "public interface", that by the way, is not complete.

Our interface is also comprised by Rego -- their builtins --, the Kubernetes version we support, the client-go we use in the controller, what policy-evaluator might read from the cluster (again, kube-rs...).

I think our dependencies and public interfaces are so many, and with many different natures impacting different kind of users (policy developers, cluster administrators...), that I don't believe semver for the toplevel versioning is meaningful.

Reply:
My main concern here is users not knowing if the upgrade of the helm charts are painless or not.
I realize now that the helm charts have their own semver, and that carries that info.

I'm happy with calver in that aspect, resolving this conversation!